src/main.rs

```diff
@@ -142,23 +156,47 @@ fn merge_sort<T: PartialOrd + std::marker::Copy + std::fmt::Debug>(v: &[T]) -> V
     return result
 }
 
+// This takes two sorted vectors, like:
+//    <5, 8, 9> and
+//    <0, 2, 3, 6>
+// and merges them into a single sorted vector like:
+//    <0, 2, 3, 5, 6, 8, 9>
+// You should be able to do this in linear time by having
+// two indices that point to where you are in xs and ys.
+// You then compare those values, push the smaller one onto
+// the result vector, and increment the appropriate index.
+// You stop when one of your indices hits the end of its
+// vector, and then push all the remaining elements from the
+// other vector onto the result.
+
+/*
+The interesting part of this method is in the for loop. We compare values in the xs subarray to values in the ys subarray
+and add the smaller one to the output vector. The condition y_pos >= ys.len() is so that we don't go out of bounds on the
+ys subarray.
+*/
 fn merge<T: PartialOrd + std::marker::Copy + std::fmt::Debug>(xs: Vec<T>, ys: Vec<T>) -> Vec<T> {
-    // This takes two sorted vectors, like:
-    //    <5, 8, 9> and
-    //    <0, 2, 3, 6>
-    // and merges them into a single sorted vector like:
-    //    <0, 2, 3, 5, 6, 8, 9>
-    // You should be able to do this in linear time by having
-    // two indices that point to where you are in xs and ys.
-    // You then compare those values, push the smaller one onto
-    // the result vector, and increment the appropriate index.
-    // You stop when one of your indices hits the end of its
-    // vector, and then push all the remaining elements from the
-    // other vector onto the result.
-
-    // This is totally wrong and will not sort. You should replace it
-    // with something useful. :)
-    return xs;
+    if xs.len() == 0 { //returning the sorted ys subarray if the xs subarray is empty
+        return ys; 
+    }
+    let mut x_pos = 0;
+    if ys.len() == 0 { //returning the sorted xs subarray if the ys subarray is empty
+        return xs;
+    }
+    let mut y_pos = 0;
+    let mut output: Vec<T> = Vec::new();
+
+    
+    for _i in 0..xs.len() + ys.len() {
+        if x_pos < xs.len() && (y_pos >= ys.len() || xs[x_pos] < ys[y_pos]) { 
```

Comment:
This conditional works, but it definitely takes some thinking about. I'd favor not special-casing the `xs.len() == 0` and `ys.len() == 0` cases above, and instead have your `for` loop just handle the cases where there is still data in both arrays. You can then have two `.extend()` one-lines to tack on any uncopied data.